cloud/google/cmd/gce-controller/cluster-controller-app/controller.go

```diff
@@ -0,0 +1,157 @@
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/cluster-api/cloud/google"
+	"sigs.k8s.io/cluster-api/cloud/google/cmd/gce-controller/cluster-controller-app/options"
+	"sigs.k8s.io/cluster-api/pkg/client/clientset_generated/clientset"
+	clusterapiclientsetscheme "sigs.k8s.io/cluster-api/pkg/client/clientset_generated/clientset/scheme"
+	"sigs.k8s.io/cluster-api/pkg/controller/cluster"
+	"sigs.k8s.io/cluster-api/pkg/controller/config"
+	"sigs.k8s.io/cluster-api/pkg/controller/sharedinformers"
+)
+
+const (
+	gceClusterControllerName = "gce-cluster-controller"
+)
+
+func StartClusterController(server *options.ClusterControllerServer, shutdown <-chan struct{}) {
+	config, err := controller.GetConfig(server.CommonConfig.Kubeconfig)
+	if err != nil {
+		glog.Fatalf("Could not create Config for talking to the apiserver: %v", err)
+	}
+
+	client, err := clientset.NewForConfig(config)
+	if err != nil {
+		glog.Fatalf("Could not create client for talking to the apiserver: %v", err)
+	}
+
+	params := google.ClusterActuatorParams{
+		ClusterClient: client.ClusterV1alpha1().Clusters(corev1.NamespaceDefault),
+	}
+	actuator, err := google.NewClusterActuator(params)
+	if err != nil {
+		glog.Fatalf("Could not create Google cluster actuator: %v", err)
+	}
+
+	si := sharedinformers.NewSharedInformers(config, shutdown)
+	// If this doesn't compile, the code generator probably
+	// overwrote the customized NewClusterController function.
+	c := cluster.NewClusterController(config, si, actuator)
+	c.RunAsync(shutdown)
+
+	select {}
+}
+
+func RunClusterController(server *options.ClusterControllerServer) error {
+	// TODO refactor to remove duplication of code between this and the machine
+	// controller server. Issue #262
+	kubeConfig, err := controller.GetConfig(server.CommonConfig.Kubeconfig)
+	if err != nil {
+		glog.Errorf("Could not create Config for talking to the apiserver: %v", err)
+		return err
+	}
+
+	kubeClientControl, err := kubernetes.NewForConfig(
+		rest.AddUserAgent(kubeConfig, "cluster-controller-manager"),
+	)
+	if err != nil {
+		glog.Errorf("Invalid API configuration for kubeconfig-control: %v", err)
+		return err
+	}
+
+	recorder, err := createRecorder(kubeClientControl)
+	if err != nil {
+		glog.Errorf("Could not create event recorder : %v", err)
+		return err
+	}
+
+	// run function will block and never return.
+	run := func(stop <-chan struct{}) {
+		StartClusterController(server, stop)
+	}
+
+	leaderElectConfig := config.GetLeaderElectionConfig()
+	if !leaderElectConfig.LeaderElect {
+		run(make(<-chan (struct{})))
+	}
+
+	// Identity used to distinguish between multiple cluster controller instances.
+	id, err := os.Hostname()
+	if err != nil {
+		return err
+	}
+
+	leaderElectionClient := kubernetes.NewForConfigOrDie(rest.AddUserAgent(kubeConfig, "cluster-leader-election"))
```

Comment:
I saw kenny's comment about the duplication -- I'm especially concerned about this section here which involves a bit of tricky leader election code that really should be shared so that both locations can get any bug fixes.

Glad we are creating an issue -- but i'm hoping it will get prioritized.

Reply:
cc @medinatiger

Reply:
I'm planning to work on it in the next couple weeks. I want to add some tests for this logic too, I think refactoring it will make that easier.

Reply:
Love the plan, +1000 on tests. Sounds good.